src/Streamly/Internal/LZ4.hs

```diff
@@ -253,45 +330,78 @@ compressChunk i ctx arr = do
-              srcLen <- coerce <$> peek hdrSrcLen
+              let hdrCompLen :: Ptr Int32 = src `plusPtr` compLenOff
+                  compData = src `plusPtr` compDataOff
+                  arrDataLen = Array.byteLength arr - metaLen
+              maxUncompLenC <- getMaxUncompLenC src
+              compLenC <- i32ToCInt <$> peek hdrCompLen
+              let compLen = cIntToInt compLenC
+                  maxCompLenC = lz4_MAX_OUTPUT_SIZE
+                  maxUncompLen = cIntToInt maxUncompLenC
 
               -- Error checks
-              if srcLen <= 0
+              if compLenC <= 0
               then error "decompressChunk: compressed data length > 2GB"
-              else if (fromIntegral :: CInt -> Int) srcLen < arrDataLen
+              else if compLen < arrDataLen
               then error $ "decompressChunk: input array data length "
                 ++ show arrDataLen ++ " is less than "
                 ++ "the compressed data length specified in the header "
-                ++ show srcLen
-              else when (srcLen > lz4_MAX_OUTPUT_SIZE) $
+                ++ show compLen
+              else when (compLenC > maxCompLenC) $
                   error $ "decompressChunk: compressed data length is more "
-                    ++ "than the max limit: " ++ show lz4_MAX_OUTPUT_SIZE
+                    ++ "than the max limit: " ++ show maxCompLenC
 
-              (MArray.Array fptr dstBegin dstMax) <-
-                  MArray.newArray ((fromIntegral :: CInt -> Int) decompLen)
-              decompLen1 <-
+              (MArray.Array fptr dstBegin dstMax)
+                  <- MArray.newArray maxUncompLen
+              decompLenC <-
                   c_decompressSafeContinue
-                        ctx compData dstBegin srcLen decompLen
-              when (decompLen1 < 0 || decompLen /= decompLen1)
+                        ctx compData dstBegin compLenC maxUncompLenC
+              when (decompLenC < 0)
                 $ error $ "decompressChunk: c_decompressSafeContinue failed. "
                     ++ "arrDataLen = " ++ show arrDataLen
-                    ++ "srcLen = " ++ show srcLen
-                    ++ "decompLen = " ++ show decompLen
-                    ++ "decompLen1 = " ++ show decompLen1
-              let dstEnd = dstBegin `plusPtr`
-                    (fromIntegral :: CInt -> Int) decompLen1
+                    ++ "compLenC = " ++ show compLenC
+                    ++ "decompLenC = " ++ show decompLenC
+              checkUncompDecompEq maxUncompLenC decompLenC
+              let decompLen = cIntToInt decompLenC
+                  dstEnd = dstBegin `plusPtr` decompLen
                   decompArr = MArray.Array fptr dstEnd dstMax
               Array.unsafeFreeze <$> MArray.shrinkToFit decompArr
 
+    where
+
+    metaLen =
+        if encodeUncompressedSize
+        then 8
+        else 4
+
+    compLenOff =
+        if encodeUncompressedSize
+        then 4
+        else 0
+
+    compDataOff =
```

Comment:
`compDataOff = metaLen`.

Reply:
Although they are equal now they have different meanings. When we have a checksum as well, `metaLen` will be `12` whereas `compDataOff` will not change.